File: problem_082.c
// -*- compile-command: "gcc -std=c99 -o problem_082 problem_082.c ceuler.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include <string.h>
#include "ceuler.h"

FILE *matrixFile;

void debugPrint(int m[80][80], int s)
{
    for (int x=0; x<s; x++)
    {
        printf("Row %2d: ", x);
        for (int y=0; y<s; y++)
            printf("%6d, ", m[x][y]);
        printf("\n");
    }
    printf("\n");
}

void computeColumn(int matrix[80][80], int size, int x, bool debug)
{
    int newCol[size];

    // Prime newCol with value + value to right
    for (int y=0; y<size; y++)
        newCol[y] = matrix[y][x] + matrix[y][x+1];

    // Get value up
    for (int y=1; y<size; y++)
    {
        int upValue = newCol[y-1] + matrix[y][x];
        if (upValue < newCol[y]) newCol[y] = upValue;
    }

    // Get value down
    for (int y=size-2; y>=0; y--)
    {
        int downValue = newCol[y+1] + matrix[y][x];
        if (downValue < newCol[y]) newCol[y] = downValue;
    }

    // Copy values back to matrix
    for (int y=0; y<size; y++)
        matrix[y][x] = newCol[y];
}

int main(){
    printf("Project Euler - Problem 82:\n"
           "Find the minimal path sum in a 31K text file containing a 80 by 80 matrix, from the top left column to right column.\n\n");

    // Begin time tracking
    struct timeval start;
    gettimeofday(&start, NULL);

    int matrix[80][80]={};
    int size = sizeof(matrix[0])/sizeof(matrix[0][0]);
    bool debug = false;

    //Populate table from text file
    matrixFile = fopen ("problem_082.txt", "rt");

    char ch;
    int x = 0, y = 0;
    while((ch = fgetc(matrixFile)))
    {
        if (ch == EOF) break;
        if (ch == ',')
        {
            x++;
            continue;
        }

        if (ch == '\n')
        {
            x = 0;
            y++;
            continue;
        }

        matrix[y][x] *= 10;
        matrix[y][x] += ch - '0';
    }

    fclose(matrixFile);

    // Solve optimal path one column at a time
    for (int x=size-2; x>=0; x--)
        computeColumn(matrix, size, x, debug);

    // Print completed matrix
    if (debug == true) debugPrint(matrix, size);

    // Find smallest value in left column
    int smallestAnswer = matrix[0][0];
    for (int y=1; y<size; y++)
        if (matrix[y][0] < smallestAnswer)
            smallestAnswer = matrix[y][0];

    printf("Smallest path sum: %d\n", smallestAnswer);
    printElapsedTime(start);
    return 0;
}

File: problem_082.rs
// -*- compile-command: "rustc -o problem_082_rs problem_082.rs" -*-
// Copyright (c) 2016 Michael Caldwell
use std::fs::File;
use std::io::Read;

fn main() {
    println!("Project Euler - Problem 82:");
    println!("Find the minimal path sum in a 31K text file containing a 80 by 80 matrix, from the left column to right column.\n");

    let mut matrix: [[u32; 80]; 80] = [[0; 80]; 80];

    // Import data from file
    let mut raw_data = String::new();
    let mut f = File::open("problem_082.txt").unwrap();
    f.read_to_string(&mut raw_data).unwrap();

    // Split data lines into vector
    let data: Vec<&str> = raw_data.split_whitespace().collect();

    // Put vector into matrix array
    for row in 0..data.len() {
        // Split line into vector of values
        let line_of_data: Vec<&str> = data[row].split(",").collect();

        for col in 0..line_of_data.len() {
            matrix[row][col] = line_of_data[col].parse::<u32>().unwrap();
        }
    }

    // Solve by columns
    for x in (0..matrix.len()-1).rev() {
        // Create a tmp column with values to the right.
        let mut tmp_col = [0; 80];
        for y in 0..matrix.len() {
            tmp_col[y] = matrix[y][x] + matrix[y][x+1];
        }

        // Solve up
        for y in 1..matrix.len() {
            let tmp_value = matrix[y][x] + tmp_col[y-1];
            if tmp_value < tmp_col[y] {
                tmp_col[y] = tmp_value;
            }
        }

        // Then solve down
        for y in (0..matrix.len()-1).rev() {
            let tmp_value = matrix[y][x] + tmp_col[y+1];
            if tmp_value < tmp_col[y] {
                tmp_col[y] = tmp_value;
            }
        }

        // Copy values back to matrix
        for y in 0..matrix.len() {
            matrix[y][x] = tmp_col[y];
        }
    }

    // Print smallest value
    let mut smallest_value = matrix[0][0];

    for y in 0..matrix.len() {
        if matrix[y][0] < smallest_value {
            smallest_value = matrix[y][0];
        }
    }

    println!("{}", smallest_value);
}
